packages/@ember/-internals/glimmer/lib/helpers/custom.ts

```diff
@@ -62,12 +73,91 @@ function hasDestroyable(manager: HelperManager): manager is HelperManagerWithDes
   return manager.capabilities.hasDestroyable;
 }
 
+let ARGS_CACHES = DEBUG ? new WeakMap<SimpleArgsProxy, Cache<Partial<Arguments>>>() : undefined;
+
+function getArgs(proxy: SimpleArgsProxy): Partial<Arguments> {
+  return getValue(DEBUG ? ARGS_CACHES!.get(proxy)! : proxy.argsCache!)!;
+}
+
+class SimpleArgsProxy {
+  argsCache?: Cache<Partial<Arguments>>;
+
+  constructor(
+    context: object,
+    computeArgs: (context: object) => Partial<Arguments> = () => EMPTY_ARGS
+  ) {
+    let argsCache = createCache(() => computeArgs(context));
+
+    if (DEBUG) {
+      ARGS_CACHES!.set(this, argsCache);
+    } else {
+      this.argsCache = argsCache;
+    }
+  }
+
+  get named() {
+    return getArgs(this).named || EMPTY_NAMED;
+  }
+
+  get positional() {
+    return getArgs(this).positional || EMPTY_POSITIONAL;
+  }
+}
+
+export function invokeHelper(
+  context: object,
+  definition: HelperDefinition,
+  computeArgs: (context: object) => Partial<Arguments>
+): Cache<unknown> {
+  assert(
+    `Expected a context object to be passed as the first parameter to invokeHelper, got ${context}`,
+    context !== null && typeof context === 'object'
+  );
+
+  const owner = getOwner(context);
+  const manager = getHelperManager(owner, definition)!;
+
+  // TODO: figure out why assert isn't using the TS assert thing
+  assert(
+    `Expected a helper definition to be passed as the second parameter to invokeHelper, but no helper manager was found. Did you use setHelperManager to associate a helper manager?`,
+    manager
+  );
+
+  let args = new SimpleArgsProxy(context, computeArgs);
+  let helper = manager.createHelper(definition, args);
```

Comment:
Lets call this `bucket` (matches what we call it in the actual `customHelper` implementation below, and helps us avoid any assumptions around what it is...)